src/middlewared/middlewared/plugins/afp.py

```diff
@@ -39,3 +42,214 @@ class Config:
+        Bool('nostat'),
+        Bool('upriv'),
+        Str('fperm'),
+        Str('dperm'),
+        Str('umask'),
+        Str('hostsallow'),
+        Str('hostsdeny'),
+        Str('auxparams'),
+        register=True
+    ))
+    async def do_create(self, data):
+        path = data['path']
+
+        await self.clean(data, 'sharingafp_create')
+        await self.validate(data, 'sharingafp_create')
+
+        if path and not os.path.exists(path):
+            try:
+                os.makedirs(path)
+            except OSError as e:
+                verrors = ValidationErrors()
+                verrors.add('sharingafp_create.path',
+                            f"Failed to create {path}: {e}")
+
+                raise verrors
+
+        data['id'] = await self.middleware.call(
+            'datastore.insert', self._config.datastore, data,
+            {'prefix': self._config.datastore_prefix})
+
+        await self.middleware.call('service.reload', 'afp')
+
+        return data
+
+    @accepts(
+        Int('id'),
+        Patch(
+            'sharingafp_create',
+            'sharingafp_update',
+            ('attr', {'update': True})
+        )
+    )
+    async def do_update(self, id, data):
+        old = await self.middleware.call(
+            'datastore.query', self._config.datastore, [('id', '=', id)],
+            {'prefix': self._config.datastore_prefix,
+             'get': True})
+
+        new = old.copy()
+        new.update(data)
+
+        await self.clean(data, 'sharingafp_update', id=id)
+        await self.validate(data, 'sharingafp_update', old=old)
+
+        await self.middleware.call(
+            'datastore.update', self._config.datastore, id, data,
+            {'prefix': self._config.datastore_prefix})
+
+        await self.middleware.call('service.reload', 'afp')
+
+        return new
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        return await self.middleware.call(
+            'datastore.delete', self._config.datastore, id)
+
+    @private
+    async def clean(self, data, schema_name, id=None):
+        clean_networks = ['hostsallow', 'hostsdeny']
+
+        for name in clean_networks:
+            data[name] = await self.clean_network(data[name], schema_name,
+                                                  name)
+
+        data['name'] = await self.name_exists(data, schema_name, id)
+
+    @private
+    async def clean_network(self, data, schema_name, attribute):
+        verrors = ValidationErrors()
+
+        if data:
+            for net in data.split(' '):
+                try:
+                    ipaddress.ip_interface(net)
+                except ValueError as e:
+                    verrors.add(f"{schema_name}.{attribute}",
+                                f'Invalid IP or Network: {net}')
+
+            if verrors:
+                raise verrors
+
+            return data.strip()
+
+        return ""
+
+    @private
+    async def validate(self, data, schema_name, old=None):
```

Comment:
It seems we could share ValidationErrors between these two calls so more errors are shown to the user at once